### PR TITLE
[X86][tablgen] Fix the broadcast tables

### PR DESCRIPTION
**Pull request by KanRobert**

### CODE REVIEW

File: llvm/utils/TableGen/X86FoldTablesEmitter.cpp [hunk 1]
```diff
@@ -529,45 +530,22 @@ void X86FoldTablesEmitter::addBroadcastEntry(
...
-    errs() << RegInstName << ", " << MemInstName << "\n";
-    llvm_unreachable("Name is not canoicalized for broadcast or "
-                     "ExeDomain is incorrect");
+  DagInit *In = MemInst->TheDef->getValueAsDag("InOperandList");
+  for (unsigned I = 0, E = In->getNumArgs(); I != E; ++I) {
+    Result.BroadcastKind =
+        StringSwitch<X86FoldTableEntry::BcastType>(In->getArg(I)->getAsString())
+            .Case("i32mem", X86FoldTableEntry::BCAST_D)
+            .Case("i64mem", X86FoldTableEntry::BCAST_Q)
+            .Case("i16mem", X86FoldTableEntry::BCAST_SH)
```
Comment (phoebewang) [inline llvm/utils/TableGen/X86FoldTablesEmitter.cpp:539]: Why not map to `BCAST_W`?

Comment (KanRobert) [reply llvm/utils/TableGen/X86FoldTablesEmitter.cpp]: Sounds good. Done.

Comment (phoebewang) [reply llvm/utils/TableGen/X86FoldTablesEmitter.cpp]: Shouldn't update here too? https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86InstrFoldTables.cpp#L300

Comment (KanRobert) [reply llvm/utils/TableGen/X86FoldTablesEmitter.cpp]: Good catch! Done.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by KanRobert** [comment]
Thanks @phoebewang

**Comment by phoebewang** [comment]
Maybe backport to LLVM18 release.

**Comment by phoebewang** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/7c3ee7cbe6419ea5e37ce2723cc1a1688380581f

**Pull request merged**